{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-processing Step\n",
    "In this step, we need to prepare the data to be ready-to-use data. This step include:\n",
    "- Preparing metadata\n",
    "- Demultiplexing FASTQ Files\n",
    "- Removing Non-Biological Sequences\n",
    "- Sequence Denoising"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xuU9p8QpDRZt"
   },
   "source": [
    "## Preparing Metadata\n",
    "Before starting the analysis, explore the sample metadata to familiarize yourself with the samples used in this study. QIIME 2 metadata is most commonly stored in a TSV (i.e. tab-separated values) file. It usually contains at least sample ID and additional information related to the samples, though metadata consisting only sample ID is still valid QIIME 2 metadata file. The instructors will guide you how to create a metadata file using Google Spreadsheet.\n",
    "\n",
    "Once you have your metadata file, you can copy it to the server. Open a new terminal/secure shell app, and run this command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "7rEw696eFmRM"
   },
   "outputs": [],
   "source": [
    "scp '/path/to/local/sample-metadata.tsv' username@remote_host:'/path/to/destination/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MUlmu8KcFpt7"
   },
   "source": [
    "Another metadata file is provided inside the materials folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IPCTgTsvGjK0"
   },
   "source": [
    "## Demultiplexing FASTQ Files\n",
    "Raw microbiome data typically exists in one of two forms: multiplexed or demultiplexed. In multiplexed data, sequences from all samples are grouped together in one or more files. In demultiplexed data, sequences are separated into different files based on the sample they are derived from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "os8Ickq5GqEe"
   },
   "outputs": [],
   "source": [
    "qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path raw-data/ \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path demux.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "H8iIW2fTS4GQ"
   },
   "source": [
    "### Summarize Demultiplexed FASTQ Files\n",
    "When you have demultiplexed sequence data, the next step is typically to generate a visual summary of it. This allows you to determine how many sequences were obtained per sample, and also to get a summary of the distribution of sequence qualities at each position in your sequence data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "cev4egxCS3g0"
   },
   "outputs": [],
   "source": [
    "qiime demux summarize \\\n",
    "  --i-data demux.qza \\\n",
    "  --o-visualization demux.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DMZiUEnWHLu0"
   },
   "source": [
    "Now, we have two types of output files:\n",
    "\n",
    "*   .qza files store QIIME 2 Artifacts on disk. QIIME 2 Artifacts represent data that are generated by QIIME 2 and intended to be used by QIIME 2, such as intermediary files in an analysis workflow. qza stands for QIIME Zipped Artifact.\n",
    "*   .qzv files store QIIME 2 Visualizations on disk. QIIME 2 Visualizations represent data that are generated by QIIME 2 and intended to be viewed by humans, such as an interactive visualization. qzv stands for QIIME Zipped Visualization.\n",
    "\n",
    ".qza and .qzv files can be loaded with [QIIME 2 View](https://view.qiime2.org/). Because we are working on our server, we need to download the files from the server before uploading it to QIIME 2 View. Open a new terminal/secure shell app and run this command."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "i2YY338WFb2f"
   },
   "outputs": [],
   "source": [
    "scp username@remote_host:'/path/to/file' '/path/to/local/destination/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "euwy-SbvWIPJ"
   },
   "source": [
    "## Removing Non-Biological Sequences\n",
    "Raw data from sequencer may contain any non-biological sequences (e.g. primers, sequencing adapters, PCR spacers, etc), we should remove these.\n",
    "\n",
    "The q2-cutadapt plugin has comprehensive methods for removing non-biological sequences from paired-end or single-end data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xBsT_YfhJltL"
   },
   "outputs": [],
   "source": [
    "qiime cutadapt trim-paired \\\n",
    "  --i-demultiplexed-sequences demux.qza \\\n",
    "  --p-front-f AGAGTTTGATCCTGGCTCAG \\\n",
    "  --p-front-r ACGGCTACCTTGTTACGACTT \\\n",
    "  --o-trimmed-sequences primer-trimmed-demux.qza \\\n",
    "  --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DRSAZrTSTLE1"
   },
   "source": [
    "## Sequence Denoising\n",
    "QIIME 2 plugins are available for several quality control methods, including DADA2, Deblur, and basic quality-score-based filtering. In this tutorial we present this step using DADA2. The result of this method will be a FeatureTable[Frequency] QIIME 2 artifact, which contains counts (frequencies) of each unique sequence in each sample in the dataset, and a FeatureData[Sequence] QIIME 2 artifact, which maps feature identifiers in the FeatureTable to the sequences they represent.\n",
    "\n",
    "The denoise-paired action, which we’ll use here, requires four parameters that are used in quality filtering:\n",
    "\n",
    "trim-left-f a, which trims off the first a bases of each forward read\n",
    "\n",
    "trunc-len-f b which truncates each forward read at position b\n",
    "\n",
    "trim-left-r c, which trims off the first c bases of each forward read\n",
    "\n",
    "trunc-len-r d which truncates each forward read at position d This allows the user to remove low quality regions of the sequences. To determine what values to pass for these parameters, you should review the Interactive Quality Plot tab in the demux.qzv file that was generated above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "_gTlD4zgTK0M"
   },
   "outputs": [],
   "source": [
    "qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs primer-trimmed-demux.qza \\\n",
    "  --p-trim-left-f 0 \\\n",
    "  --p-trunc-len-f 270 \\\n",
    "  --p-trim-left-r 0 \\\n",
    "  --p-trunc-len-r 200 \\\n",
    "  --o-representative-sequences rep-seqs.qza \\\n",
    "  --o-table asv-table.qza \\\n",
    "  --o-denoising-stats stats.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JxOY8d7LT4aJ"
   },
   "source": [
    "After DADA2 completes, you’ll want to explore the resulting data. We can do this using the following commands, which will create visual summaries of the data. The feature-table summarize action command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "MAhMwoM6T4G-"
   },
   "outputs": [],
   "source": [
    "qiime feature-table summarize \\\n",
    "  --i-table asv-table.qza \\\n",
    "  --o-visualization asv-table.qzv \\\n",
    "  --m-sample-metadata-file sample-metadata.tsv\n",
    "\n",
    "qiime feature-table tabulate-seqs \\\n",
    "  --i-data rep-seqs.qza \\\n",
    "  --o-visualization rep-seqs.qzv\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file stats.qza \\\n",
    "  --o-visualization stats.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PRfW0HgkUBrQ"
   },
   "source": [
    "If the sequences appear to have low quality score, we can filter them further using the `q2-quality-filter` plugin and run the . For example, we want to filter out the sequences that have qscore below 28, we can run this command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "HY35RgPwUiuu"
   },
   "outputs": [],
   "source": [
    "qiime quality-filter q-score \\\n",
    "  --i-demux primer-trimmed-demux.qza \\\n",
    "  --p-min-quality 28 \\\n",
    "  --o-filtered-sequences demux-filtered.qza \\\n",
    "  --o-filter-stats demux-filter-stats.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Congratulations, you have sucessfully generate ready-to-use data. Now continue to the next section of the tutorial: [Microbiome Classification Analysis](4_Microbiome_Classification_Analysis.ipynb),"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or you just can back to the main tutorial page [Main Page](1_Metagenomics_Workshop_Module.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
